File: backend/app/api/scrape.py
from uuid import uuid4
from datetime import datetime, timedelta

from fastapi import APIRouter, BackgroundTasks, HTTPException

from app.db.database import SessionLocal
from app.db.repository import get_latest_job, upsert_job
from app.core.config import get_settings
from app.models.schemas import (
    ManualIngestRequest,
    ManualIngestResponse,
    ScrapeRequest,
    ScrapeResponse,
    StatusResponse,
)
from app.pipeline.chunk_embed import rebuild_faiss_index, save_raw_documents
from app.pipeline.scrape_apify import scrape_urls

router = APIRouter(prefix="/scrape", tags=["scrape"])
STALE_JOB_TIMEOUT = timedelta(minutes=15)


def _run_scrape_job(job_id: str, urls: list[str]) -> None:
    settings = get_settings()
    with SessionLocal() as db:
        upsert_job(db, job_id=job_id, status="running", message="Scraping started")

    try:
        documents = scrape_urls(settings=settings, urls=urls)
        total_docs = save_raw_documents(settings, documents)
        build_result = rebuild_faiss_index(settings)

        with SessionLocal() as db:
            upsert_job(
                db,
                job_id=job_id,
                status="completed",
                message="Scrape and indexing completed",
                scraped_documents=total_docs,
                indexed_chunks=build_result.get("chunks", 0),
            )
    except Exception as exc:
        with SessionLocal() as db:
            upsert_job(db, job_id=job_id, status="failed", message=str(exc))


@router.post("", response_model=ScrapeResponse)
def scrape(request: ScrapeRequest, background_tasks: BackgroundTasks) -> ScrapeResponse:
    settings = get_settings()

    if not settings.apify_api_token:
        raise HTTPException(status_code=500, detail="APIFY_API_TOKEN is not configured")
    if not settings.active_api_key:
        raise HTTPException(status_code=500, detail="Active LLM API key is not configured")

    job_id = str(uuid4())
    with SessionLocal() as db:
        upsert_job(db, job_id=job_id, status="queued", message=f"Job {job_id} queued")

    background_tasks.add_task(_run_scrape_job, job_id, request.urls)

    return ScrapeResponse(job_id=job_id, status="queued")


@router.post("/manual", response_model=ManualIngestResponse)
def manual_ingest(request: ManualIngestRequest) -> ManualIngestResponse:
    settings = get_settings()

    documents = [item.model_dump() for item in request.documents]
    saved = save_raw_documents(settings, documents)

    indexed_chunks = 0
    status = "completed"
    message = f"Saved {saved} document(s)"

    if request.reindex:
        if not settings.active_api_key:
            raise HTTPException(status_code=500, detail="Active LLM API key is not configured")

        try:
            result = rebuild_faiss_index(settings)
            indexed_chunks = int(result.get("chunks", 0))
            message = f"Saved {saved} document(s) and reindexed corpus"
        except Exception as exc:
            status = "partial"
            message = f"Saved {saved} document(s), but reindex failed: {exc}"

    return ManualIngestResponse(
        status=status,
        message=message,
        saved_documents=saved,
        indexed_chunks=indexed_chunks,
    )


@router.post("/reindex", response_model=StatusResponse)
def reindex() -> StatusResponse:
    settings = get_settings()

    if not settings.active_api_key:
        raise HTTPException(status_code=500, detail="Active LLM API key is not configured")

    job_id = str(uuid4())
    with SessionLocal() as db:
        upsert_job(db, job_id=job_id, status="running", message="Reindex started")

    try:
        result = rebuild_faiss_index(settings)
    except Exception as exc:
        with SessionLocal() as db:
            upsert_job(db, job_id=job_id, status="failed", message=f"Reindex failed: {exc}")
        raise HTTPException(status_code=500, detail=f"Reindex failed: {exc}")

    with SessionLocal() as db:
        upsert_job(
            db,
            job_id=job_id,
            status="completed",
            message="Manual reindex completed",
            scraped_documents=result.get("documents", 0),
            indexed_chunks=result.get("chunks", 0),
        )

    return StatusResponse(
        status="completed",
        message="Manual reindex completed",
        scraped_documents=result.get("documents", 0),
        indexed_chunks=result.get("chunks", 0),
    )


@router.get("/status", response_model=StatusResponse)
def status() -> StatusResponse:
    try:
        with SessionLocal() as db:
            job = get_latest_job(db)

            if job is not None and job.status == "running":
                updated_at = job.updated_at or job.created_at
                if updated_at and datetime.utcnow() - updated_at > STALE_JOB_TIMEOUT:
                    job = upsert_job(
                        db,
                        job_id=job.job_id,
                        status="failed",
                        message="Scrape timed out before completion",
                        scraped_documents=job.scraped_documents,
                        indexed_chunks=job.indexed_chunks,
                    )
    except Exception as exc:
        raise HTTPException(status_code=503, detail=f"Database unavailable: {exc}")

    if job is None:
        return StatusResponse(status="idle", message="Ready", scraped_documents=0, indexed_chunks=0)

    return StatusResponse(
        status=job.status,
        message=job.message,
        scraped_documents=job.scraped_documents,
        indexed_chunks=job.indexed_chunks,
    )

File: backend/src/api/scrape.ts
import { Router, Request, Response } from "express";
import { v4 as uuidv4 } from "uuid";
import {
  ScrapeRequest,
  ManualIngestRequest,
  validateScrapeRequest,
  validateManualIngestRequest,
} from "../models/schemas";
import { scrapeUrls } from "../pipeline/scrape_apify";
import {
  saveRawDocuments,
  rebuildFaissIndex,
} from "../pipeline/chunk_embed";
import { upsertJob, getLatestJob } from "../db/repository";
import { getSettings } from "../core/config";
import logger from "../core/logger";

export const scrapeRouter = Router();

const STALE_JOB_TIMEOUT = 15 * 60 * 1000; // 15 minutes

async function runScrapeJob(jobId: string, urls: string[]): Promise<void> {
  const settings = getSettings();

  try {
    await upsertJob(jobId, "running", "Scraping started");

    const documents = await scrapeUrls(urls);
    const totalDocs = await saveRawDocuments(documents);

    logger.info(`Scraped ${totalDocs} documents`);

    const buildResult = await rebuildFaissIndex();

    await upsertJob(
      jobId,
      "completed",
      "Scrape and indexing completed",
      totalDocs,
      buildResult.chunks
    );
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger.error(`Scrape job ${jobId} failed:`, error);
    await upsertJob(jobId, "failed", message);
  }
}

scrapeRouter.post("", async (req: Request, res: Response) => {
  try {
    const validated = validateScrapeRequest(req.body);
    const settings = getSettings();

    if (!settings.apifyApiToken) {
      return res
        .status(500)
        .json({ detail: "APIFY_API_TOKEN is not configured" });
    }
    if (!settings.getActiveApiKey()) {
      return res
        .status(500)
        .json({ detail: "Active LLM API key is not configured" });
    }

    const jobId = uuidv4();
    await upsertJob(jobId, "queued", `Job ${jobId} queued`);

    // Run scrape job in background
    runScrapeJob(jobId, validated.urls).catch((error) => {
      logger.error(`Background scrape job failed:`, error);
    });

    res.json({ job_id: jobId, status: "queued" });
  } catch (error) {
    logger.error("Scrape request validation failed:", error);
    const message = error instanceof Error ? error.message : String(error);
    res.status(400).json({ detail: message });
  }
});

scrapeRouter.post("/manual", async (req: Request, res: Response) => {
  try {
    const validated = validateManualIngestRequest(req.body);
    const settings = getSettings();

    const documents = validated.documents.map((item) => ({
      source_url: item.source_url,
      title: item.title,
      text: item.text,
    }));

    const saved = await saveRawDocuments(documents);

    let indexed = 0;
    let status = "completed";
    let message = `Saved ${saved} document(s)`;

    if (validated.reindex) {
      if (!settings.getActiveApiKey()) {
        return res
          .status(500)
          .json({ detail: "Active LLM API key is not configured" });
      }

      try {
        const buildResult = await rebuildFaissIndex();
        indexed = buildResult.chunks;
        message = `Saved ${saved} document(s), indexed ${indexed} chunk(s)`;
      } catch (error) {
        status = "error";
        message = `Saved but indexing failed: ${error}`;
        logger.error("Reindexing failed:", error);
      }
    }

    res.json({
      status,
      message,
      saved_documents: saved,
      indexed_chunks: indexed,
    });
  } catch (error) {
    logger.error("Manual ingest request failed:", error);
    const message = error instanceof Error ? error.message : String(error);
    res.status(400).json({ detail: message });
  }
});

scrapeRouter.get("/status", async (req: Request, res: Response) => {
  try {
    const job = await getLatestJob();

    if (!job) {
      return res.json({
        status: "idle",
        message: "No scrape job has been started yet",
        scraped_documents: 0,
        indexed_chunks: 0,
      });
    }

    // Check if job is stale
    const updatedAt = job.updatedAt || new Date();
    const isStale = Date.now() - updatedAt.getTime() > STALE_JOB_TIMEOUT;

    if (isStale && job.status === "running") {
      await upsertJob(
        job.jobId,
        "stalled",
        "Job appears to be stalled"
      );
      job.status = "stalled";
    }

    res.json({
      status: job.status,
      message: job.message,
      scraped_documents: job.scrapedDocuments,
      indexed_chunks: job.indexedChunks,
    });
  } catch (error) {
    logger.error("Status request failed:", error);
    res.status(500).json({ detail: "Failed to get status" });
  }
});
